submariner/fine-grained-connectivity.md

```diff
@@ -0,0 +1,86 @@
+A Clusterset is a group of clusters that forms a trust boundary. The design of Clusterset can be aligned
+to the MCS KEP. More details are available in the below link.
+
+[Cluster-Id KEP]: http://www.reddit.comhttps://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/2149-clusterid
+
+In this proposal, we intend to use only the ClusterId(which exists already) and ClusterSet as a part of
+Submariner Cluster CRD. The ClusterClaim API shall be realized once the KEP is finalized upstream.
+
+### Design Details
+
+![Clustersets](./images/clustersets.png)
+
+The Submariner should be enhanced to support grouping of clusters into Clustersets. The Cluster CRD can
+be enhanced to hold the Clustersets name in which the cluster is a part of. The one or more Clustersets
+the cluster wants to be a part of shall be specified when the cluster joins the broker. If no
+Clustersets are specified, the default will be used. The cluster-id should be unique across Clustersets.
+
+```Go
+
+type Cluster struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec              ClusterSpec `json:"spec"`
+}
+
+type ClusterSpec struct {
+        ClusterID   string   `json:"cluster_id"` // perhaps this could just be a hash of the name...?
+        ClusterSets []string `json:"cluster_sets"`
+        ServiceCIDR []string `json:"service_cidr"`
+        ClusterCIDR []string `json:"cluster_cidr"`
+        GlobalCIDR  []string `json:"global_cidr"`
+}
+```
+
+#### Submariner Broker
+
+There will be a single broker running in one of the participating clusters. The clusters will specify
+the Clusterset it wants to be a part of when it joins the broker. A cluster can be a part of multiple Clustersets.
+The broker ensures that resources are only synced within a Clusterset.
+
+#### Service Discovery
+
+The broker ensures that the ServiceImports are synced only within the Clustersets. So a cluster should be a
+part of the Clusterset from which the service was exported to receive the service import.  But the same
+service in the same namespace across Clustersets would be considered an instance the same service.
+
+#### Dataplane Connectivity
+
+Similar to service imports endpoint objects are synced only within the Clustersets. A full mesh of tunnels
+will be created within the members of the Clustersets. If a cluster is a part of two Clusterset multiple
+tunnels will be created.
+
+#### GlobalNet
+
+Global net currently allocates CIDR per cluster. So across Clustersets each cluster will have non-overlapping CIDR.
+
+## Alternatives
+
+One broker per Clusterset: With multiple brokers either we need multiple Submariner components each having its own
+namespace and talk to a specific broker or we need to have a submariner component talking to multiple brokers.
+Either option requires changes across Submariner and Lighthouse. A significant change will be required in the
+submariner-operator and deployment as well.
```

Comment:
Either way, there would be broker and Submariner components per cluster set, whether they run in the same process/pod or separate ones. Of course there's pros and cons to each,.

Reply:
Right. But there is a problem if multiple Submariner Components are run (even if they are in different namespace) since the Submariner Pods run with HostNetworking, they modify the routingRules/ipTables/Tunnels etc on the Host There will be race conditions if multiple Submariner Pods are run. So IMHO its better to avoid that path and use single instance of Submariner Pods even if the Cluster is part of multiple Clustersets and use Broker for intelligently distributing endpoint/clusters based on the Clusterset.

Reply:
As discussed we will be going with one broker and one broker n/s for now.